### ISSUE TITLE
Make table copyable for Sheets/Excel

### ISSUE THREAD

**Issue by atishpatel** [vaadin/vaadin-grid#848]
Right now, when I select rows and columns in vaadin-grid and paste them into Google Sheets, it just becomes a long column. If I was to use the <table> tag or use div tags with `display: table-row` and `display: table-cell` and "paste values only" it into Google Sheets, the column and row are pasted properly. 

If I change the CSS `.vaadin-grid-table-row` to `display: table-row` and `.vaadin-grid-table-cell` to `display: table-cell`, the copy-paste functionality works, but it messes up the look of everything because vaadin-grid relies on flex. Is there a way to work around this?

P.S. This element is super awesome so thank you!

**Comment by Saulis** [comment]
Which version and browser are you running, I'm able to copy-paste from https://cdn.vaadin.com/vaadin-grid/2.0.0/demo/ with Chrome to Google Sheets?

**Comment by atishpatel** [comment]
The latest from branch 2.1-polymer-2. I just tested, and it works with the link you gave me (version 2.0). But, it doesn't work on 2.1. :\

**Comment by atishpatel** [comment]
I just looked at it for a little bit. It looks like in 2.0 there is `td is="vaadin-grid-table-cell"` while 2.1 uses divs or vaadin-grid-table-cell. O_o I'm using shadow dom instead of shady dom. I don't know if that makes a difference.

**Comment by tomivirkki** [comment]
This is not something that the `<vaadin-grid>` officially supports. Copy-pasting to sheets happens to work with Chrome due to the grid’s internal table structure, and as you discovered, in `<vaadin-grid> 2.0` we used extended native elements (`td is="vaadin-grid-table-cell"`) internally which made the DOM structure possible in the first place.

But since Polymer 2 [no longer supports Type-extension elements](https://www.polymer-project.org/2.0/docs/about_20#type-extension), we need to drop the table structure in the hybrid version and revert to using normal custom elements internally.

**Comment by atishpatel** [comment]
The type-extension isn't currently supported, but it might be in the future. https://github.com/Polymer/polymer/issues/3925 

It would be a nice feature to have in vaadin-grid 2.1+ if possible because it makes the element almost useless for me otherwise. I'm not familiar with this so bare with me. :)

Would doing this work?
`class vaadin-grid-table-row extends HTMLTableRowElement`

If not, I guess we can wait until there is a resolution on this issue. https://github.com/w3c/webcomponents/issues/509

Thank you!

**Comment by jouni** [comment]
It is possible to work around this limitation even when using non-table elements. We could listen copy events and populate the clipboard data manually. We quickly prototyped this witj @saulis sole tile ago.

**Comment by atishpatel** [comment]
@jouni Is the prototype in a local branch? If so, can you please push it upstream?

**Comment by jouni** [comment]
@atishpatel I remember us doing it adhoc in jsbin, not on top of vaadin-grid. We just prototyped the generic concept of intercepting copy events and how to populate the clipboard data, which could be used in vaadin-grid.

**Comment by davidmaxwaterman** [comment]
an aside: is it possible to paste *from* Excel/etc into vaadin-grid?

**Comment by web-padawan** [comment]
The latest 4.0.0 stable version has table structure re-implemented under the hood. I tried copying rows from the [demo](https://cdn.vaadin.com/vaadin-grid/4.0.0/demo/#grid-basic-demos) and it works with both Google Sheets and Excel (tested in MS Office 2007). Therefore I am closing this issue.

**Comment by limonte** [comment]
> The latest 4.0.0 stable version has table structure re-implemented under the hood. I tried copying rows from the demo and it works with both Google Sheets and Excel (tested in MS Office 2007). Therefore I am closing this issue.

Still not quite right:

![recorded](https://user-images.githubusercontent.com/6059356/33022673-9d2caa34-ce0e-11e7-836c-4b758e257fe3.gif)

Expected would be: when pasting from the gif above, rows from 0 to 21 inclusively are pasted.

**Comment by jouni** [comment]
Yeah, you can only copy the rows/cells that are actually visible in the DOM.

**Comment by tomivirkki** [comment]
> This is not something that the <vaadin-grid> officially supports

This might currently work (on Chrome) to some extent but all our official features should work in all supported browsers. I think the only way to make this work nicely across browsers is to implement virtual selection and populating the clipboard programmatically.
